File: Python/Permutations II.py
class Solution:
    # @param num, a list of integer
    # @return a list of lists of integers
    def permuteUnique(self, num):
        num.sort()
        return self.helper(num)
        
    def helper(self, num):
        if len(num) == 1: return [num]
        ret = []
        for index, cur in enumerate(num):
            if index > 0 and num[index-1] == cur:
                continue
            print self.helper(num[:index] + num[index+1:])
            ret += [[cur] + prev for prev in self.helper(num[:index] + num[index+1:])]
            print ret
        return ret
                
        
s = Solution()
print s.permuteUnique([1,1,2])        

File: Java/src/Permutations II.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution
{
  public List<List<Integer>> permuteUnique(int[] num)
  {
    Arrays.sort(num);
    List<Integer> intList = new ArrayList<Integer>();
    for (int index = 0; index < num.length; index++)
    {
      intList.add(num[index]);
    }
    return helper(intList);

  }

  public List<List<Integer>> helper(List<Integer> intList)
  {
    List<List<Integer>> ret = new ArrayList<>();
    if (intList.size() == 1)
    {
      ret.add(intList);
      return ret;
    }

    for (int index = 0; index < intList.size(); index++)
    {
      if (index > 0 && intList.get(index) == intList.get(index - 1))
        continue;
      List<Integer> prevList = new ArrayList<>(intList);
      prevList.remove(index);
      List<List<Integer>> prevRet = helper(prevList);
      for (List<Integer> l : prevRet)
      {
        l.add(0, intList.get(index));
        ret.add(l);
      }
    }
    return ret;
  }
}
